### ISSUE TITLE
persistence: scale-out testing 

### ISSUE THREAD

**Issue by ruchirK** [MaterializeInc/materialize#8463]
Mostly will be musing, and reports of what we find as we test materialize with persistent user tables and persistent kafka upsert sources.

Relevant general questions:

* How good is the observed write/read throughput compared to the expected max write/read throughput of your disk?

Dimensions to test:

* Total number of records sent to Materialize
* Number of records inserted / second
* Working set size (distinct number of un-deleted key-val pairs at any time)
* Number of workers
* Number of streams
* (for kafka upsert sources) number of kafka partitions
* timestamping frequency?
* compaction enabled / disabled / varying compaction lag

General thoughts:

- tables don't let us perform deletes at the moment, but we can simulate the effect of a delete by sending duplicate records that get consolidated down - in a delete two records turn into 0 over time, in this case 2 (or N) records turn into 1 over time
- perhaps we can simulate the pipelining of writes in sources by sending many records at a time in tables

**Comment by ruchirK** [comment]
Learnings from the past day:

Have been focused mostly on testing with user tables.

The first performance bottleneck we ran into is that a very large `VALUES` clause (e.g. more than 20K records) turns into a very large union expression that takes very long to plan and optimize (even though it ultimately reduces to a constant expression). Something like https://github.com/MaterializeInc/materialize/pull/8477 would help a lot with that.

In the meantime, I've been unblocked thanks to the recent work that allows for `INSERT` statements into a table to be populated by the results of `SELECT`s from other tables. This lets me do something like:

```
materialize=> create table ten (f1 int);
CREATE TABLE
materialize=> create table thousand (f1 int);
CREATE TABLE
materialize=> create table to_insert(k1 int, k2 int, v text);
CREATE TABLE
materialize=> create table t1 (k1 int, k2 int, v text);
CREATE TABLE
materialize=> insert into ten values (0), (1), (2), (3), (4);
INSERT 0 5
materialize=> insert into ten values (5), (6), (7), (8), (9);
INSERT 0 5
materialize=> insert into thousand select a1.f1 * 100 + a2.f1 * 10 + a3.f1 from ten as a1, ten as a2, ten as a3;
INSERT 0 1000
materialize=> insert into to_insert select a1.f1, a2.f1, 'aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa' from thousand as a1, thousand as a2;
INSERT 0 1000000
materialize=> insert into t1 select * from to_insert;
INSERT 0 1000000
```

and then repeatedly invoke `insert into t1 select * from to_insert;` to insert 1 million ~110 byte records into `t1`, which generates a ~170 MB unsealed batch.

Sidenote: 170 MB seemed high to me because the row consists of 2 integers + a 100 byte string so conservatively I would expect something like say 4 + 4 + 100 + 8 (for length of the string) + 8 (misc type info metadata) so 124 bytes, but according to `mz_row_size()` it's actually 144 bytes. 

```
materialize=> select mz_row_size(to_insert.*) from to_insert limit 1;
 mz_row_size
-------------
         144
(1 row)
```
The `pg_column` sizes seem more in line with what i would expect

```
materialize=> select pg_column_size(to_insert.*) from to_insert limit 1;
 pg_column_size
----------------
            121
(1 row)

Time: 85.478 ms
materialize=> select pg_column_size(to_insert.v) from to_insert limit 1;
 pg_column_size
----------------
            102
(1 row)

Time: 101.345 ms
materialize=> select pg_column_size(to_insert.k1) from to_insert limit 1;
 pg_column_size
----------------
              5
(1 row)

Time: 110.030 ms
materialize=> select pg_column_size(to_insert.k2) from to_insert limit 1;
 pg_column_size
----------------
              5
(1 row)

Time: 125.925 ms
materialize=> select pg_column_size((to_insert.k1, to_insert.k2)) from to_insert limit 1;
 pg_column_size
----------------
             19
(1 row)

Time: 52.249 ms
```

Anyway, that's something to investigate for later. 

The actual work that the coordinator / persistence runtime are doing here now takes about ~1.1 seconds when compaction and writing to unsealed are disabled. (The end-to end query time takes about 2.5 seconds but it takes roughly 1.5 seconds to perform the `SELECT`. )

The bottlenecks for this 1.1 second of work right now are:

1. The coordinator spends ~200 ms allocating vectors and filling them with encoded bytes in `write_atomic`. Possible hypothesis here that packing all of the encoded bytes in one array might reduce cache misses and make this faster.
2. The persistence runtime spends ~200 ms cloning hashtables in `drain_pending`, and doing allocations / other work in `set_unsealed_batch`.
3. There's roughly a ~170 - 200 ms gap where neither the persistence runtime nor the coordinator thread are doing any CPU intensive work. Best guess for that at the moment is that's how long it takes to write the file down.
4. The persistence runtime spends another ~400 - 450 ms performing listen callbacks (~200 ms) and clearing out hashtables (~100 ms) and cloning vectors (~100 ms)

I'm going to focus my energies this morning on reducing the costs of all of these hashtable and vector clones.

**Comment by ruchirK** [comment]
Theres two other large clones that drive the fast path time up.

1. We make a clone before calling `set_unsealed_batch` so that we can keep the `BlobUnsealedBatch` in the read cache. This isn't strictly necessary to be done in the fast path, and removing this moves the test time to < 0.7 seconds.

2. We clone (and decode, and do various other things) to send writes to downstream listeners. This work can also be done on a separate thread away from the main fast path.

Aside from this, the only other obvious bottleneck in the fast path at the moment is the time spent allocating thousands of small vectors in `write_atomic`. @danhhz is working on an api that can let us expose a columnar representation that doesn't need to be composed of many small allocations (and potentially can also factorize the inputs so that repeated keys / values are only stored once).

I think that if we address 1 and 2 fast path time for this test should be < 0.5 seconds

**Comment by ruchirK** [comment]
I'm going to switch gears a little bit, and move to investigating maintenance workloads to try to see whats the largest working set of data we currently support. The persistence work for me breaks down into a write fast (latency sensitive) path, a read fast (latency sensitive) path, and a maintenance slow (throughput sensitive path). I think, though am not sure the main things to look at here are:

1. How much CPU / RAM utilization does background work drive for a given size working set? E.g. 1 GB of distinct 100byte records? 10 GB?
2. How much disk utilization does background work drive?

**Comment by danhhz** [comment]
Before you switch into the compaction work, let's investigate whether we can sustain the write performance you've gotten us to. So, we can write 1 million updates in ~.8 - 1 second, can we write 60 million in 1m? 30x that in 30m?

**Comment by ruchirK** [comment]
oh thats a good question -- the numbers will be a little bit lower because the way I'm testing Materialize right now by doing inserts like

```
insert into t1 select * from to_insert;
```

takes about 1 - 1.5 seconds to process the `select *` query. I don't know a more performant way to insert 1 MM rows offhand, but I can poke around to see

**Comment by ruchirK** [comment]
I realized something this morning that made me go down a bit of a rabbit hole.

The current tests we are doing are nice and show us inserting 100 mb of data in 2 seconds end to end, but its a bit of a false image because we aren't really inserting new data - we're just seeding a table with data from another table, which itself was constructed by multiplying out a very small amount of data sent to a user table. This doesn't actually give us an accurate picture for what end-to-end performance a user should expect when using persisted tables. 

When I try to simulate a more realistic workload with `COPY ... FROM STDIN` which seems like the fastest way to get data into mz it seems to take approximately 11 seconds

```
altaria-2:~ Test$ time cat inserts.data | psql -h localhost -p 6875 -U materialize -c "copy t1 from stdin with(delimiter=',')"
COPY 1000000

real	0m11.160s
user	0m0.109s
sys	0m0.088s
```

most of which is spent in optimization. I'm following up with the sql / coord teams because fixing that is outside of my area of expertise, and kind of outside of the scope of this issue.

**Comment by ruchirK** [comment]
https://github.com/MaterializeInc/materialize/pull/8609 improves the time spent optimizing COPY FROM .. by a factor of 2 by improving the way we do constant folding

**Comment by ruchirK** [comment]
closing this issue out as we've made a lot more improvements to persist latency, and I'm focusing on persist's memory usage on initial ingestion and restart, which will be covered in another issue
